File: app/src/server/convert/inline.ts
type Annotations = {
  bold?: boolean;
  italic?: boolean;
  strikethrough?: boolean;
  underline?: boolean;
  code?: boolean;
};

interface Segment {
  type?: string;
  plain_text?: string;
  href?: string | null;
  annotations?: Annotations;
  equation?: { expression: string };
  mention?: { type: string; page?: { id: string }; database?: { id: string } };
}

export function richTextToMd(segments: Segment[]): string {
  return segments.map(segmentToMd).join('');
}

function segmentToMd(seg: Segment): string {
  if (seg.type === 'equation' && seg.equation) return `$${seg.equation.expression}$`;
  if (seg.type === 'mention') return mentionToMd(seg);
  return textToMd(seg);
}

function textToMd(seg: Segment): string {
  const text = seg.plain_text ?? '';
  if (!text) return '';
  const a = seg.annotations ?? {};
  let out = text;
  if (a.code) out = `\`${out}\``;
  if (a.bold && a.italic) out = `***${out}***`;
  else if (a.bold) out = `**${out}**`;
  else if (a.italic) out = `*${out}*`;
  if (a.strikethrough) out = `~~${out}~~`;
  if (a.underline) out = `<u>${out}</u>`;
  if (seg.href) out = `[${out}](${seg.href})`;
  return out;
}

function mentionToMd(seg: Segment): string {
  const m = seg.mention;
  const plain = seg.plain_text ?? '';
  if (!m) return plain;
  if (m.type === 'page' && m.page) return `{{notion-link:${m.page.id}|${plain}}}`;
  if (m.type === 'database' && m.database) return `{{notion-link:${m.database.id}|${plain}}}`;
  return plain;
}

File: src/notion_extractor/convert/inline.py
from typing import Any


def rich_text_to_md(segments: list[dict[str, Any]]) -> str:
    return "".join(_segment_to_md(s) for s in segments)


def _segment_to_md(seg: dict[str, Any]) -> str:
    seg_type = seg.get("type", "text")

    if seg_type == "equation":
        return f"${seg['equation']['expression']}$"

    if seg_type == "mention":
        return _mention_to_md(seg)

    return _text_to_md(seg)


def _text_to_md(seg: dict[str, Any]) -> str:
    text = seg.get("plain_text", "")
    if not text:
        return ""

    ann = seg.get("annotations", {})
    href = seg.get("href")

    formatted = text
    if ann.get("code"):
        formatted = f"`{formatted}`"
    if ann.get("bold") and ann.get("italic"):
        formatted = f"***{formatted}***"
    elif ann.get("bold"):
        formatted = f"**{formatted}**"
    elif ann.get("italic"):
        formatted = f"*{formatted}*"
    if ann.get("strikethrough"):
        formatted = f"~~{formatted}~~"
    if ann.get("underline"):
        formatted = f"<u>{formatted}</u>"
    if href:
        formatted = f"[{formatted}]({href})"
    return formatted


def _mention_to_md(seg: dict[str, Any]) -> str:
    mention = seg.get("mention", {})
    m_type = mention.get("type")
    plain = seg.get("plain_text", "")

    if m_type == "page":
        page_id = mention["page"]["id"]
        return f"{{{{notion-link:{page_id}|{plain}}}}}"
    if m_type == "database":
        db_id = mention["database"]["id"]
        return f"{{{{notion-link:{db_id}|{plain}}}}}"
    if m_type == "date":
        return plain
    if m_type == "user":
        return plain
    return plain
